packages/admin-ui-core/src/styleVariants.tsx

```diff
@@ -0,0 +1,51 @@
+import { get } from '@vtex/admin-ui-util'
+
+import type { StyleProp } from './types'
+
+export function style(csx: StyleProp): StyleProp {
+  return csx
+}
+
+export function styleVariants<V extends {}>(variants: VariantDef<V>) {
+  return function variantFn(rules: VariantsCall<V>) {
+    const css: StyleProp = {}
+
+    for (const key in rules) {
+      Object.assign(css, get(variants, `${key}.${rules[key]}`, {}))
+    }
+
+    return style(css)
+  }
+}
+
+export type VariantDef<V> = {
+  [k in keyof V]: { [b in keyof V[k]]: StyleProp }
+}
+
+export type InferVariant<T> = T extends 'true'
+  ? boolean
+  : T extends 'false'
+  ? boolean
+  : T
```

Comment:
Suggestion: Could it be like this?
```suggestion
export type InferVariant<T> = (T extends 'true' || T extends 'false')
  ? boolean
  : T
```